subprojects/gradle/runner/service/src/main/kotlin/com/avito/runner/service/listener/TestMetricsSender.kt

```diff
@@ -0,0 +1,66 @@
+package com.avito.runner.service.listener
+
+import com.avito.android.stats.GaugeMetric
+import com.avito.android.stats.StatsDSender
+import com.avito.logger.LoggerFactory
+import com.avito.logger.create
+import com.avito.runner.service.model.TestCase
+import com.avito.runner.service.model.TestCaseRun
+import com.avito.runner.service.worker.device.Device
+import com.avito.time.TimeProvider
+
+class TestMetricsSender(
+    private val statsDSender: StatsDSender,
+    private val timeProvider: TimeProvider,
+    loggerFactory: LoggerFactory
+) : TestListener {
+
+    private val prefix = "runner.test.duration"
+
+    private val logger = loggerFactory.create<TestMetricsSender>()
+
+    private val intendedMarks = mutableMapOf<Key, Long>()
+
+    private val startedMarks = mutableMapOf<Key, Long>()
+
+    override fun intended(test: TestCase, targetPackage: String, executionNumber: Int) {
+        intendedMarks[Key(test, executionNumber)] = timeProvider.nowInMillis()
+    }
+
+    override fun started(device: Device, targetPackage: String, test: TestCase, executionNumber: Int) {
+        val key = Key(test, executionNumber)
+        startedMarks[key] = timeProvider.nowInMillis()
+
+        val intendedMark = intendedMarks[key]
+
+        if (intendedMark != null) {
+            statsDSender.send(prefix, GaugeMetric("queue", timeProvider.nowInMillis() - intendedMark))
+        } else {
+            logger.warn("Can't send test metric: intended mark not found")
+        }
+    }
+
+    override fun finished(
+        device: Device,
+        test: TestCase,
+        targetPackage: String,
+        result: TestCaseRun.Result,
+        durationMilliseconds: Long,
+        executionNumber: Int
+    ) {
+        val key = Key(test, executionNumber)
+
+        val startedMark = startedMarks[key]
+
+        if (startedMark != null) {
+            statsDSender.send(prefix, GaugeMetric("execution", timeProvider.nowInMillis() - startedMark))
```

Comment:
How many unique metrics and events on each PR do you expect?
I'm worried about the workload.

Reply:
it's 2 metrics for each test run (with retries), so about 6000 per PR; seems ok